File: sem1/t4_5/Stack.cpp
#include <iostream>
#include "Stack.h"

struct StackElement
{
    StackElement *next;
    char value;
};

struct Stack
{
    StackElement *top;
    int size;
};


Stack *createStack()
{
    Stack *newStack = new Stack;
    newStack->top = nullptr;
    newStack->size = 0;
    return newStack;
}

void push(Stack *s, char value)
{
    StackElement *elem = new StackElement;
    elem->next = s->top;
    elem->value = value;
    s->top = elem;
    s->size++;
}

bool isEmpty(Stack const *s)
{
    return s->size == 0;
}

char pop(Stack *s)
{
    if (isEmpty(s))
        return 'e';

    char value = s->top->value;

    StackElement *toDelete = s->top;
    s->top = s->top->next;
    s->size--;
    delete toDelete;

    return value;
}

char top(Stack *s)
{
    if (isEmpty(s))
        return 'e';
    else
        return s->top->value;
}

void clearStack(Stack *s)
{
    if (isEmpty(s))
        return;

    StackElement *temp = s->top;
    s->top = nullptr;
    s->size = 0;

    while (temp->next != nullptr)
    {
        StackElement *toDelete = temp;
        temp = temp->next;
        delete toDelete;
    }

    delete temp;
}

File: course_1/semester_2/hw1/task_1/src/puzyrev/sem2/hw1/task1/Stack.java
package puzyrev.sem2.hw1.task1;

/**
 * Stack data structure.
 * @param <Type> type of stack element
 */
public class Stack<Type> {

    /**
     * Stack element class.
     */
    private class StackElement {
        private Type value;
        private StackElement next;

        public StackElement(Type value, StackElement next) {
            this.value = value;
            this.next = next;
        }

        public Type getValue() {
            return value;
        }

        public StackElement getNext() {
            return next;
        }
    }

    private StackElement head = null;

    /**
     * Add element to top.
     * @param value element to be adding
     */
    public void push(Type value) {
        head = new StackElement(value, head);
    }

    /**
     * Get top element and remove it from Stack.
     * @return top element
     */
    public Type pop() {
        if (head == null) {
            return null;
        } else {
            Type value = head.getValue();
            head = head.getNext();
            return value;
        }
    }

    /**
     * Check if Stack is empty.
     * @return 'true' if Stack is empty or 'false' if not.
     */
    public boolean isEmpty() {
        return head == null;
    }
}